This bug used to exist upstream as well, but was fixed when the
VLAN code path was restructured in 2010.

The approach here is different from upstream in order to minimise
the effect on the code.

net: Fix memory leak/corruption on VLAN GRO_DROP

The function napi_reuse_skb is only meant to be used for packets
merged by GRO.  Using it on the VLAN path will lead to memory
leaks/corruption.  This patch is based on Jay Vosburgh's patch,
and it fixes the problem by calling kfree_skb on the VLAN GRO_DROP
path instead of napi_reuse_skb.

Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
[dannf: backported to Debian's 2.6.32]

diff -urpN linux-source-2.6.32.orig/net/core/dev.c linux-source-2.6.32/net/core/dev.c
--- linux-source-2.6.32.orig/net/core/dev.c	2011-06-11 13:10:41.000000000 -0600
+++ linux-source-2.6.32/net/core/dev.c	2011-08-03 22:42:37.774288259 -0600
@@ -2650,6 +2650,9 @@ gro_result_t napi_frags_finish(struct na
 		break;
 
 	case GRO_DROP:
+		kfree_skb(skb);
+		break;
+
 	case GRO_MERGED_FREE:
 		napi_reuse_skb(napi, skb);
 		break;
